### PR TITLE
fix(jetbrains): improve diff performance and session layout

### PR DESCRIPTION
**Pull request by kirillk**
## Context

This improves the JetBrains plugin experience for large branch diffs, long session transcripts, and queued-prompt UX. The main problems were expensive full-transcript restyling while rendering diff previews, large inline patches freezing the EDT, cramped diff file names, existing sessions sometimes loading with cropped transcript content until the panel was resized, and the prompt send/stop button getting out of sync when attachments changed during a busy session.

## Implementation

- Cap large inline diff previews and show an overflow placeholder instead of rendering huge patches directly in chat tool cards.
- Avoid session-wide markdown restyling fan-out when individual transcript blocks are styled.
- Improve JetBrains diff editor/tree rendering for large branch diffs, including compact relative paths and horizontal scrolling.
- Add a bounded transcript reflow after existing-session load/restyle so width-keyed cached heights are refreshed after HTML panes settle, without disabling the validate-root/cache optimization.
- Keep the busy-session prompt send/stop button synchronized when attachments are added, removed, or cleared.
- Fix layered overlay preferred sizing to avoid double-scaling under IDE zoom.

### Cap preview size so it won't slow down UI

Screen Shot 2026-07-31 at 3 48 41 PM

### Manual/local verification

- Agent ran `./gradlew typecheck` from `packages/kilo-jetbrains/`.
- Agent ran `./gradlew :frontend:test --tests "ai.kilocode.client.session.ui.PromptPanelTest"` from `packages/kilo-jetbrains/`.
- Agent ran `./gradlew :frontend:test --tests "ai.kilocode.client.session.ui.SessionLayoutTest" --tests "ai.kilocode.client.session.ui.SessionMessageListPanelTest" --tests "ai.kilocode.client.session.ui.SessionRootPanelTest"` from `packages/kilo-jetbrains/`.
- Agent ran `./gradlew :frontend:test --tests "ai.kilocode.client.session.SessionUiLayoutTest" --tests "ai.kilocode.client.session.SessionScrollTest"` from `packages/kilo-jetbrains/`.
- Earlier commits in this branch include JetBrains diff-focused frontend/backend regression tests for diff rendering, overflow capping, and modified-files behavior.

### Reviewer test steps

1. Open the JetBrains plugin with a branch containing a large diff.
2. Confirm the diff tree uses compact relative paths and can scroll horizontally for long paths.
3. Trigger/read tool output containing a very large patch and confirm the chat card shows an overflow placeholder instead of freezing the EDT.
4. Open an existing long chat session and confirm the transcript is fully laid out on first load without needing to resize the tool window.
5. Switch editor theme or font settings and confirm transcript content reflows correctly without the previous session-wide freeze.
6. While a session is busy with an empty prompt, add and remove an attachment; confirm the prompt button switches between Send when a draft exists and Stop when no draft remains.

### Blocked checks and substitute verification

- No blocked checks. The relevant JetBrains typecheck and targeted frontend tests passed locally.

## Checklist

- [x] Issue linked above, or exception explained
- [x] Tests/verification described
- [x] Screenshots/video included for visual changes, or marked N/A
- [x] Changeset considered for user-facing changes
- [x] I personally reviewed the diff and can explain the changes, including any AI-assisted work.

## Get in Touch

Kilo maintainers can reach me through the usual GitHub PR thread.

### CODE REVIEW

**Review by lambertjosh — APPROVED** [review]

File: packages/kilo-jetbrains/frontend/src/main/kotlin/ai/kilocode/client/session/ui/SessionMessageListPanel.kt [hunk 1]
```diff
@@ -473,6 +501,34 @@ class SessionMessageListPanel(
...
+        if (dead || id != seq) return
+        if (turnViews.isEmpty()) return
+        val changed = reflow()
+        if (changed) onReflow?.invoke(true)
+        if (remaining <= 0) {
+            stable = -1
+            return
+        }
+        val height = preferredSize.height
+        val left = if (height == stable) remaining - 1 else REFLOW_PASSES
```
[the remark this answers is not in the document]
Comment (kirillk) [reply packages/kilo-jetbrains/frontend/src/main/kotlin/ai/kilocode/client/session/ui/SessionMessageListPanel.kt]: Fixed in 820d5d5. `reflowPass` now carries a hard total-pass budget (`REFLOW_BUDGET = REFLOW_PASSES * 4`) that decrements on every pass and never resets, alongside the existing `remaining` counter that restarts on height change. A streaming session whose height changes each EDT cycle can no longer reset the budget forever and hold the panel in a perpetual `forgetAll()`/re-measure loop — the chain always terminates within a bounded number of passes while still settling across a few height changes as HTML panes reflow. Went with the hard-budget option rather than bailing while busy, since a session opened mid-stream still needs the crop-on-load reflow. Covered by a new test that latches `pendingReflow` via a real turn at zero width.

File: packages/kilo-jetbrains/backend/src/main/kotlin/ai/kilocode/backend/rpc/KiloWorkspaceRpcApiImpl.kt [hunk 2]
```diff
@@ -81,6 +81,7 @@ class KiloWorkspaceRpcApiImpl internal constructor(
         private val GLOBAL = MODERN + LEGACY + "config.json"
         private val LOCAL_DIRS = listOf(".kilo", ".kilocode", ".opencode")
         private const val DIFF_CAP = 200_000
+        private const val BRANCH_DIFF_CAP = 8 * 1024 * 1024
```
[the remark this answers is not in the document]
Comment (kirillk) [reply packages/kilo-jetbrains/backend/src/main/kotlin/ai/kilocode/backend/rpc/KiloWorkspaceRpcApiImpl.kt:84]: Keeping the eager cap for now. The 200 KB -> 8 MB bump was deliberate: at 200 KB a normal branch diff blanked most files' patches, which is the regression this PR set out to fix. The payload is still bounded — `capDiff` stops fetching once the running total reaches `BRANCH_DIFF_CAP` and after `MAX_OVERSIZED_PATCHES` oversized misses — and the frontend already refuses to render patches over `DIFF_MAX_LINES` inline.

On-demand per-file patching is a reasonable improvement, but it's a larger change to the `branchDiff` RPC contract and the diff editor's tree-selection flow than fits this PR (fetch-on-select, request coalescing, refresh invalidation, tests). Tracking it as a follow-up rather than reworking the contract here.

File: packages/kilo-jetbrains/frontend/src/test/kotlin/ai/kilocode/client/session/ui/SessionMessageListPanelTest.kt [hunk 3]
```diff
@@ -144,6 +144,51 @@ class SessionMessageListPanelTest : BasePlatformTestCase() {
...
+        val child = Growing(20)
+        panel.add(child, 0)
+        panel.setSize(0, 400)
+        layout(panel)
+        child.markValid()
+        child.size = 80
+
+        // At zero width an HTML pane collapses to a one-char column, so reflow must report no
+        // change instead of locking the transcript height in against that bogus measurement.
+        assertFalse(panel.reflow())
```
[the remark this answers is not in the document]
Comment (kirillk) [reply packages/kilo-jetbrains/frontend/src/test/kotlin/ai/kilocode/client/session/ui/SessionMessageListPanelTest.kt:170]: Done in 820d5d5. Added `test deferred reflow re-arms on first real width layout`: it seeds a real turn (`model.upsertMessage(msg("u1", "user"))`) so a zero-width `reflow()` actually latches `pendingReflow`, then the first real-width `doLayout()` consumes it and schedules a pass. The child is left valid at the same width with a changed intrinsic size, so a plain layout keeps the cached height and only the re-armed `forgetAll()` reflow re-measures it — the assertion fails if the `doLayout()` re-arm branch is removed. The existing turnless test still covers the zero-width skip.

File: packages/kilo-jetbrains/frontend/src/main/kotlin/ai/kilocode/client/diff/KiloDiffEditorContent.kt [hunk 4]
```diff
@@ -584,9 +623,12 @@ private class Renderer : JPanel(BorderLayout()), TreeCellRenderer {
         val name = item?.name?.ifBlank { item.path }.orEmpty()
         val color = item?.file?.let(::fileStatus)?.color
         if (color == null) text.append(name) else text.append(name, SimpleTextAttributes(SimpleTextAttributes.STYLE_PLAIN, color))
-        val changed = item != null && (item.additions != 0 || item.deletions != 0)
-        badge.isVisible = changed
-        if (changed) badge.update(item.additions, item.deletions)
+        // A folder's badge rolls up its descendants' stats, which is only meaningful while the
+        // folder is collapsed. Once expanded the child rows carry their own badges, so hide the
+        // folder aggregate to avoid duplicating the numbers. Leaf files always show their badge.
+        val show = item != null && (item.additions != 0 || item.deletions != 0) && !(item.dir && expanded)
```
[the remark this answers is not in the document]
Comment (kirillk) [reply packages/kilo-jetbrains/frontend/src/main/kotlin/ai/kilocode/client/diff/KiloDiffEditorContent.kt:664]: Good catch — fixed in 87b00a9. `buildFileTree` now installs a `TreeExpansionListener` that calls `TreeUtil.invalidateCacheAndRepaint(tree.ui)` on expand and collapse, so the folder row re-measures immediately instead of keeping its stale expanded-state bounds when the badge comes back.

Heads-up: `TreeUtil.invalidateCacheAndRepaint` is `@ApiStatus.Experimental`, so this leans on an experimental IntelliJ API.

On testing: as you noted, the painted-width regression only shows on a live tree that caches path bounds across expand/collapse — the headless harness re-measures `getPathBounds` on demand, so it can't reproduce the stale cache. I added `test folder row width tracks its badge visibility`, which asserts the folder row's measured width grows when collapsed; it guards the expansion-dependent width that motivates the invalidation, and I've left it honestly scoped to that rather than claiming it exercises the cache behavior.

File: packages/kilo-jetbrains/frontend/src/main/kotlin/ai/kilocode/client/session/ui/SessionMessageListPanel.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (kirillk) [reply packages/kilo-jetbrains/frontend/src/main/kotlin/ai/kilocode/client/session/ui/SessionMessageListPanel.kt]: Good point — fixed in 45ed8f2 by gating on the condition the cap actually targets, per your first option. `reflowPass` now only treats a moving height as an unsettled layout while the session is idle:

```kotlin
val left = if (height == stable || model.state.isBusy()) remaining - 1 else REFLOW_PASSES
```

While busy, height changes are just streamed content, so the pass counts down and the chain settles in `REFLOW_PASSES` and hands off to the per-turn `forgetTurn` path — that removes the runaway at its source rather than relying on the raw count. While idle a moving height genuinely means panes are still settling, so `remaining` keeps restarting and the chain terminates by stabilizing, which is the common crop-on-load case; it won't trip the budget mid-settle there. `budget` stays purely as the hard backstop against a pathological idle pane that never converges, so exhaustion no longer coincides with the normal settling path.

File: packages/kilo-jetbrains/frontend/src/test/kotlin/ai/kilocode/client/session/ui/SessionMessageListPanelTest.kt (hunk 3 above)
[the remark this answers is not in the document]
Comment (kirillk) [reply packages/kilo-jetbrains/frontend/src/test/kotlin/ai/kilocode/client/session/ui/SessionMessageListPanelTest.kt:170]: Added in 45ed8f2, along the lines you suggested. `test reflow budget terminates when height never settles` adds an `EverGrowing` child whose `getPreferredSize()` returns a taller height on every call, so the idle chain restarts its settle window on every pass and never stabilizes. `UIUtil.dispatchAllInvocationEvents()` would spin forever without the cap; with it the drain returns and the pass count stays bounded (measured 25, i.e. the `REFLOW_PASSES * 4` budget region — the assertion caps it at 30). A future change that reset `budget` alongside `remaining` would hang the drain or blow the bound.

File: packages/kilo-jetbrains/frontend/src/main/kotlin/ai/kilocode/client/session/ui/SessionMessageListPanel.kt [hunk 5]
```diff
@@ -473,6 +520,48 @@ class SessionMessageListPanel(
...
+            return
+        }
+        val changed = reflow()
+        if (changed) onReflow?.invoke(true)
+        // [remaining] restarts whenever the height is still moving, so the chain keeps re-measuring
+        // until the layout settles for REFLOW_PASSES consecutive passes. [budget] never resets: a
+        // session that keeps streaming changes its height every EDT cycle, which would otherwise
+        // reset [remaining] forever and hold the panel in a perpetual forgetAll()/re-measure loop.
+        // The hard budget caps total passes so streaming can't defeat the width-keyed height cache.
+        if (remaining <= 0 || budget <= 0) {
```
[the remark this answers is not in the document]
Comment (kirillk) [reply packages/kilo-jetbrains/frontend/src/main/kotlin/ai/kilocode/client/session/ui/SessionMessageListPanel.kt:551]: Good point — fixed in 45ed8f2 by gating on the condition the cap actually targets, per your first option. `reflowPass` now only treats a moving height as an unsettled layout while the session is idle:

```kotlin
val left = if (height == stable || model.state.isBusy()) remaining - 1 else REFLOW_PASSES
```

While busy, height changes are just streamed content, so the pass counts down and the chain settles in `REFLOW_PASSES` and hands off to the per-turn `forgetTurn` path — that removes the runaway at its source rather than relying on the raw count. While idle a moving height genuinely means panes are still settling, so `remaining` keeps restarting and the chain terminates by stabilizing, which is the common crop-on-load case; it won't trip the budget mid-settle there. `budget` stays purely as the hard backstop against a pathological idle pane that never converges, so exhaustion no longer coincides with the normal settling path.

File: packages/kilo-jetbrains/frontend/src/test/kotlin/ai/kilocode/client/session/ui/SessionMessageListPanelTest.kt [hunk 6]
```diff
@@ -144,6 +144,78 @@ class SessionMessageListPanelTest : BasePlatformTestCase() {
...
+
+        // Once the panel is laid out with a real width the child measures to its true height.
+        panel.setSize(600, 400)
+        layout(panel)
+        assertEquals(80, child.height)
+    }
+
+    fun `test deferred reflow re-arms on first real width layout`() {
```
[the remark this answers is not in the document]
Comment (kirillk) [reply packages/kilo-jetbrains/frontend/src/test/kotlin/ai/kilocode/client/session/ui/SessionMessageListPanelTest.kt:178]: Added in 45ed8f2, along the lines you suggested. `test reflow budget terminates when height never settles` adds an `EverGrowing` child whose `getPreferredSize()` returns a taller height on every call, so the idle chain restarts its settle window on every pass and never stabilizes. `UIUtil.dispatchAllInvocationEvents()` would spin forever without the cap; with it the drain returns and the pass count stays bounded (measured 25, i.e. the `REFLOW_PASSES * 4` budget region — the assertion caps it at 30). A future change that reset `budget` alongside `remaining` would hang the drain or blow the bound.

File: packages/kilo-jetbrains/frontend/src/main/kotlin/ai/kilocode/client/diff/KiloDiffEditorContent.kt [hunk 7]
```diff
@@ -421,17 +442,47 @@ private fun buildFileTree(files: List<DiffFileDto>): Tree {
         val node = path.lastPathComponent as? DefaultMutableTreeNode
         (node?.userObject as? Node)?.name.orEmpty()
     }
+    // A folder row hides its rolled-up badge while expanded, so its preferred width depends on
+    // expansion state. JTree only invalidates cached path bounds on model changes, not on
+    // expand/collapse, so a collapsed folder would keep its narrower expanded-state bounds and the
+    // re-shown badge would squeeze the name until an unrelated re-measure. Invalidate the layout
+    // cache on toggle so the row re-measures immediately.
+    tree.addTreeExpansionListener(object : TreeExpansionListener {
```
[the remark this answers is not in the document]
Comment (kirillk) [reply packages/kilo-jetbrains/frontend/src/main/kotlin/ai/kilocode/client/diff/KiloDiffEditorContent.kt:454]: Right, that would have made the exact large diffs this PR targets quadratic. Fixed in 6820530.

- The toggle listener now skips invalidation while a bulk op is running (`if (!tree.bulk) ...`).
- `expandAll`/`collapseAll` go through a `bulkToggle(tree) { ... }` helper that sets the flag for the loop and calls `invalidateCacheAndRepaint` exactly once at the end, so each bulk op is O(rows) again.
- The listener is registered *after* the initial `expandAll(tree)` in `buildFileTree` (already fully expanded, nothing to re-measure), and `applyFiles` still invalidates via the preceding `tree.model = buildFileModel(...)`.

Per-row user toggles — the case the fix is for — still invalidate immediately.

File: packages/kilo-jetbrains/frontend/src/main/kotlin/ai/kilocode/client/session/ui/SessionMessageListPanel.kt [hunk 8]
```diff
@@ -473,6 +520,51 @@ class SessionMessageListPanel(
...
+            stable = -1
+            return
+        }
+        val height = preferredSize.height
+        // A moving height means the layout is still settling only while the session is idle. During
+        // streaming it just tracks incoming content, so restarting the settle window on every delta
+        // was the runaway that pinned the panel in a perpetual forgetAll()/re-measure loop. Count the
+        // pass down while busy so streaming settles in REFLOW_PASSES and hands off to the per-turn
+        // forgetTurn path; [budget] still caps the idle case if a pane never converges.
+        val left = if (height == stable || model.state.isBusy()) remaining - 1 else REFLOW_PASSES
```
[the remark this answers is not in the document]
Comment (kirillk) [reply packages/kilo-jetbrains/frontend/src/main/kotlin/ai/kilocode/client/session/ui/SessionMessageListPanel.kt]: You're right — `isBusy()` was too broad, and the `recoverPending()` path makes it a real regression. Fixed in f194769 by gating on the state that actually streams:

```kotlin
val left = if (height == stable || model.state is SessionState.Busy) remaining - 1 else REFLOW_PASSES
```

Now only `SessionState.Busy` (deltas landing every EDT cycle) shortens the settle window; awaiting-permission/question, retry, and offline keep restarting on a moving height, so a session opened blocked on a permission prompt still settles the transcript instead of stopping after `REFLOW_PASSES`. The runaway fix stays because the streaming case is exactly `Busy`.

File: packages/kilo-jetbrains/frontend/src/test/kotlin/ai/kilocode/client/session/ui/SessionMessageListPanelTest.kt [hunk 9]
```diff
@@ -144,6 +144,97 @@ class SessionMessageListPanelTest : BasePlatformTestCase() {
...
+        model.loadHistory(listOf(MessageWithPartsDto(msg("u1", "user"), emptyList())))
+        panel.add(EverGrowing(), 0)
+        panel.setSize(600, 400)
+
+        UIUtil.dispatchAllInvocationEvents()
+
+        // Reaching this line proves the chain terminated. The pass count is bounded by the hard
+        // budget (REFLOW_PASSES * 4), so a regression that reset it alongside `remaining` would
+        // either hang here or blow past this bound.
+        assertTrue("reflow passes must be bounded by the budget, was $reflows", reflows in 1..30)
```
[the remark this answers is not in the document]
Comment (kirillk) [reply packages/kilo-jetbrains/frontend/src/test/kotlin/ai/kilocode/client/session/ui/SessionMessageListPanelTest.kt:221]: Added in f194769. `test streaming session settles reflow within the pass window` reuses the `EverGrowing` child but sets `SessionState.Busy("thinking")` before draining the EDT. The chain now stops after ~`REFLOW_PASSES` (measured 7 passes) instead of draining the idle budget (~25), and the assertion caps it at 10 — dropping or inverting the `is SessionState.Busy` term restarts the window every pass and blows past that bound. The existing idle budget test still covers the `Idle` path.

File: packages/kilo-jetbrains/frontend/src/test/kotlin/ai/kilocode/client/session/ui/SessionMessageListPanelTest.kt [hunk 10]
```diff
@@ -144,6 +144,117 @@ class SessionMessageListPanelTest : BasePlatformTestCase() {
...
+    }
+
+    fun `test streaming session settles reflow within the pass window`() {
```
[the remark this answers is not in the document]
Comment (kirillk) [reply packages/kilo-jetbrains/frontend/src/test/kotlin/ai/kilocode/client/session/ui/SessionMessageListPanelTest.kt:224]: Good point — neither existing test distinguished the predicate. Added in b2f17a2: `test non-streaming active state keeps the reflow settle window` uses the same `EverGrowing` child but sets `SessionState.Retry("retrying", 1, 0L)` — `isBusy() == true` but not `SessionState.Busy`. It asserts the pass count reaches the idle budget region (`11..30`), i.e. the window keeps restarting rather than collapsing to `REFLOW_PASSES`.

Verified it locks in the fix: reverting the gate to `isBusy()` drops it to 7 passes and the test fails. That's the awaiting-permission/retry/offline case `recoverPending()` can seed after load.
